cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -0,0 +1,151 @@
+                        {"iso_year": 1899, "iso_week": 52, "day_of_week": 7},
+                        {"iso_year": 2033, "iso_week": 20, "day_of_week": 3}, null])");
+  auto quarter = "[1, 1, 1, 2, null]";
+  auto hour = "[0, 23, 0, 3, null]";
+  auto minute = "[0, 23, 59, 33, null]";
+  auto second = "[59.123456789, 23.999999999, 20.001001001, 20.0, null]";
+  auto millisecond = "[123, 999, 1, 0, null]";
+  auto microsecond = "[456, 999, 1, 0, null]";
+  auto nanosecond = "[789, 999, 1, 0, null]";
+  auto subsecond = "[123456789, 999999999, 1001001, 0, null]";
+
+  CheckScalarUnary("year", unit, times, int64(), year);
+  CheckScalarUnary("month", unit, times, int64(), month);
+  CheckScalarUnary("day", unit, times, int64(), day);
+  CheckScalarUnary("day_of_week", unit, times, int64(), day_of_week);
+  CheckScalarUnary("day_of_year", unit, times, int64(), day_of_year);
+  CheckScalarUnary("iso_year", unit, times, int64(), iso_year);
+  CheckScalarUnary("iso_week", unit, times, int64(), iso_week);
+  CheckScalarUnary("iso_calendar", ArrayFromJSON(unit, times), iso_calendar);
+  CheckScalarUnary("quarter", unit, times, int64(), quarter);
+  CheckScalarUnary("hour", unit, times, int64(), hour);
+  CheckScalarUnary("minute", unit, times, int64(), minute);
+  CheckScalarUnary("second", unit, times, float64(), second);
+  CheckScalarUnary("millisecond", unit, times, int64(), millisecond);
+  CheckScalarUnary("microsecond", unit, times, int64(), microsecond);
+  CheckScalarUnary("nanosecond", unit, times, int64(), nanosecond);
+  CheckScalarUnary("subsecond", unit, times, int64(), subsecond);
+}
+
+TEST(ScalarTemporalTest, TestTemporalComponentExtractionWithDifferentUnits) {
+  auto iso_calendar_type =
+      struct_({field("iso_year", int64()), field("iso_week", int64()),
+               field("day_of_week", int64())});
+  const char* times =
+      R"(["1970-01-01T00:00:59","2000-02-29T23:23:23",
+            "1899-01-01T00:59:20","2033-05-18T03:33:20", null])";
+  auto year = "[1970, 2000, 1899, 2033, null]";
+  auto month = "[1, 2, 1, 5, null]";
+  auto day = "[1, 29, 1, 18, null]";
+  auto day_of_week = "[4, 2, 7, 3, null]";
+  auto day_of_year = "[1, 60, 1, 138, null]";
+  auto iso_year = "[1970, 2000, 1899, 2033, null]";
+  auto iso_week = "[1, 9, 52, 20, null]";
+  auto iso_calendar =
+      ArrayFromJSON(iso_calendar_type,
+                    R"([{"iso_year": 1970, "iso_week": 1, "day_of_week": 4},
+                          {"iso_year": 2000, "iso_week": 9, "day_of_week": 2},
+                          {"iso_year": 1899, "iso_week": 52, "day_of_week": 7},
+                          {"iso_year": 2033, "iso_week": 20, "day_of_week": 3}, null])");
+  auto quarter = "[1, 1, 1, 2, null]";
+  auto hour = "[0, 23, 0, 3, null]";
+  auto minute = "[0, 23, 59, 33, null]";
+  auto second = "[59, 23, 20, 20, null]";
+
+  //  for (auto u : internal::AllTimeUnits()) {
+  for (auto u : {TimeUnit::NANO}) {
```

Comment:
This is not yet testing the different units?

Reply:
As stated above:
> This test fails for every unit except for nanosecond even though we [register kernels for all units](https://github.com/apache/arrow/pull/10176/commits/efd6b9e746dadaaaa6d1cc69614e6145b30b8037#diff-53850537a2f3f7977473b30241fdd3ddb35aa97269284c06471d9b0c599e2c59R415).
Is this expected? We can get timestamp units at runtime but matching would be preferable I think.

Reply:
Ah, sorry, missed that comment.